File: src/photonic_qc/quantum_state.py
"""
Quantum State Representation for Photonic Quantum Computing

This module provides the PhotonicState class for representing and manipulating
quantum states of photons in a quantum computer.
"""

import numpy as np
from typing import List, Tuple, Optional


class PhotonicState:
    """
    Represents a quantum state of photonic qubits.
    
    The state is represented as a complex-valued vector in the computational basis.
    For n qubits, the state vector has dimension 2^n.
    
    Attributes:
        state_vector (np.ndarray): Complex vector representing the quantum state
        num_qubits (int): Number of qubits in the system
    """
    
    def __init__(self, state_vector: np.ndarray, normalize: bool = True):
        """
        Initialize a photonic quantum state.
        
        Args:
            state_vector: Complex vector representing the quantum state
            normalize: Whether to normalize the state vector (default: True)
        """
        self.state_vector = np.array(state_vector, dtype=complex)
        
        # Determine number of qubits
        size = len(self.state_vector)
        self.num_qubits = int(np.log2(size))
        
        if 2 ** self.num_qubits != size:
            raise ValueError(f"State vector size must be a power of 2, got {size}")
        
        if normalize:
            self.normalize()
    
    def normalize(self):
        """Normalize the state vector to unit length."""
        norm = np.linalg.norm(self.state_vector)
        if norm > 1e-10:
            self.state_vector /= norm
    
    def copy(self) -> 'PhotonicState':
        """Create a copy of the state."""
        return PhotonicState(self.state_vector.copy(), normalize=False)
    
    @classmethod
    def zero_state(cls, num_qubits: int) -> 'PhotonicState':
        """
        Create a |0...0⟩ state with n qubits.
        
        Args:
            num_qubits: Number of qubits
            
        Returns:
            PhotonicState in |0...0⟩ state
        """
        size = 2 ** num_qubits
        state_vector = np.zeros(size, dtype=complex)
        state_vector[0] = 1.0
        return cls(state_vector, normalize=False)
    
    @classmethod
    def one_state(cls, num_qubits: int) -> 'PhotonicState':
        """
        Create a |1...1⟩ state with n qubits.
        
        Args:
            num_qubits: Number of qubits
            
        Returns:
            PhotonicState in |1...1⟩ state
        """
        size = 2 ** num_qubits
        state_vector = np.zeros(size, dtype=complex)
        state_vector[-1] = 1.0
        return cls(state_vector, normalize=False)
    
    @classmethod
    def basis_state(cls, num_qubits: int, basis_index: int) -> 'PhotonicState':
        """
        Create a computational basis state |i⟩.
        
        Args:
            num_qubits: Number of qubits
            basis_index: Index of the basis state (0 to 2^n - 1)
            
        Returns:
            PhotonicState in basis state |i⟩
        """
        size = 2 ** num_qubits
        if basis_index < 0 or basis_index >= size:
            raise ValueError(f"Basis index must be between 0 and {size-1}")
        
        state_vector = np.zeros(size, dtype=complex)
        state_vector[basis_index] = 1.0
        return cls(state_vector, normalize=False)
    
    @classmethod
    def superposition(cls, num_qubits: int) -> 'PhotonicState':
        """
        Create an equal superposition state (|0⟩ + |1⟩)/√2 ⊗ ... ⊗ (|0⟩ + |1⟩)/√2.
        
        Args:
            num_qubits: Number of qubits
            
        Returns:
            PhotonicState in equal superposition
        """
        size = 2 ** num_qubits
        state_vector = np.ones(size, dtype=complex) / np.sqrt(size)
        return cls(state_vector, normalize=False)
    
    def density_matrix(self) -> np.ndarray:
        """
        Compute the density matrix ρ = |ψ⟩⟨ψ| for this pure state.
        
        Returns:
            Density matrix as a 2D numpy array
        """
        return np.outer(self.state_vector, self.state_vector.conj())
    
    def is_normalized(self, tolerance: float = 1e-10) -> bool:
        """
        Check if the state is normalized.
        
        Args:
            tolerance: Numerical tolerance for the norm check
            
        Returns:
            True if ||ψ|| ≈ 1
        """
        norm = np.linalg.norm(self.state_vector)
        return abs(norm - 1.0) < tolerance
    
    def probability(self, basis_index: int) -> float:
        """
        Compute the probability of measuring a specific basis state.
        
        Args:
            basis_index: Index of the basis state
            
        Returns:
            Probability P(|i⟩) = |⟨i|ψ⟩|²
        """
        if basis_index < 0 or basis_index >= len(self.state_vector):
            raise ValueError(f"Basis index must be between 0 and {len(self.state_vector)-1}")
        
        return abs(self.state_vector[basis_index]) ** 2
    
    def probabilities(self) -> np.ndarray:
        """
        Compute probabilities for all basis states.
        
        Returns:
            Array of probabilities for each basis state
        """
        return np.abs(self.state_vector) ** 2
    
    def inner_product(self, other: 'PhotonicState') -> complex:
        """
        Compute inner product ⟨φ|ψ⟩ with another state.
        
        Args:
            other: Another PhotonicState
            
        Returns:
            Complex inner product
        """
        if self.num_qubits != other.num_qubits:
            raise ValueError("States must have the same number of qubits")
        
        return np.vdot(other.state_vector, self.state_vector)
    
    def fidelity(self, other: 'PhotonicState') -> float:
        """
        Compute fidelity |⟨φ|ψ⟩|² with another state.
        
        Args:
            other: Another PhotonicState
            
        Returns:
            Fidelity value between 0 and 1
        """
        return abs(self.inner_product(other)) ** 2
    
    def expectation_value(self, operator: np.ndarray) -> complex:
        """
        Compute expectation value ⟨ψ|O|ψ⟩ of an operator.
        
        Args:
            operator: Operator matrix
            
        Returns:
            Expectation value
        """
        return np.vdot(self.state_vector, operator @ self.state_vector)
    
    def partial_trace(self, traced_qubits: List[int]) -> np.ndarray:
        """
        Compute partial trace over specified qubits.
        
        Args:
            traced_qubits: List of qubit indices to trace out
            
        Returns:
            Reduced density matrix
        """
        # Get full density matrix
        rho = self.density_matrix()
        
        # Determine kept qubits
        kept_qubits = [i for i in range(self.num_qubits) if i not in traced_qubits]
        
        if not kept_qubits:
            # All qubits traced out, return scalar
            return np.array([[np.trace(rho)]])
        
        # Dimensions
        dim_kept = 2 ** len(kept_qubits)
        dim_traced = 2 ** len(traced_qubits)
        
        # Initialize reduced density matrix
        rho_reduced = np.zeros((dim_kept, dim_kept), dtype=complex)
        
        # Sum over traced qubits
        for i in range(dim_kept):
            for j in range(dim_kept):
                # Map reduced indices to full indices
                for k in range(dim_traced):
                    # Build full indices
                    idx_i = 0
                    idx_j = 0
                    
                    # Insert kept qubit values
                    for pos, qubit in enumerate(kept_qubits):
                        bit_i = (i >> pos) & 1
                        bit_j = (j >> pos) & 1
                        idx_i |= (bit_i << qubit)
                        idx_j |= (bit_j << qubit)
                    
                    # Insert traced qubit values (same for both)
                    for pos, qubit in enumerate(traced_qubits):
                        bit = (k >> pos) & 1
                        idx_i |= (bit << qubit)
                        idx_j |= (bit << qubit)
                    
                    rho_reduced[i, j] += rho[idx_i, idx_j]
        
        return rho_reduced
    
    def __str__(self) -> str:
        """String representation of the state."""
        result = []
        for i, amplitude in enumerate(self.state_vector):
            if abs(amplitude) > 1e-10:
                basis = format(i, f'0{self.num_qubits}b')
                if amplitude.imag == 0:
                    result.append(f"{amplitude.real:+.4f}|{basis}⟩")
                else:
                    result.append(f"({amplitude.real:+.4f}{amplitude.imag:+.4f}j)|{basis}⟩")
        return " + ".join(result) if result else "0"
    
    def __repr__(self) -> str:
        """Representation of the state."""
        return f"PhotonicState(num_qubits={self.num_qubits}, state={str(self)})"

File: src/PhotonicQuantumComputer/QuantumState.cs
using System.Numerics;
using System.Text;
using MathNet.Numerics.LinearAlgebra;
using MathNet.Numerics.LinearAlgebra.Complex;

namespace PhotonicQuantumComputer;

/// <summary>
/// Represents a quantum state of photonic qubits.
/// The state is represented as a complex-valued vector in the computational basis.
/// For n qubits, the state vector has dimension 2^n.
/// </summary>
public class PhotonicState
{
    /// <summary>
    /// Complex vector representing the quantum state
    /// </summary>
    public Complex[] StateVector { get; private set; }

    /// <summary>
    /// Number of qubits in the system
    /// </summary>
    public int NumQubits { get; private set; }

    /// <summary>
    /// Initialize a photonic quantum state.
    /// </summary>
    /// <param name="stateVector">Complex vector representing the quantum state</param>
    /// <param name="normalize">Whether to normalize the state vector (default: true)</param>
    public PhotonicState(Complex[] stateVector, bool normalize = true)
    {
        StateVector = (Complex[])stateVector.Clone();

        // Determine number of qubits
        int size = StateVector.Length;
        NumQubits = (int)Math.Log2(size);

        if ((1 << NumQubits) != size)
        {
            throw new ArgumentException($"State vector size must be a power of 2, got {size}");
        }

        if (normalize)
        {
            Normalize();
        }
    }

    /// <summary>
    /// Normalize the state vector to unit length.
    /// </summary>
    public void Normalize()
    {
        double norm = Math.Sqrt(StateVector.Sum(c => c.Real * c.Real + c.Imaginary * c.Imaginary));
        if (norm > 1e-10)
        {
            for (int i = 0; i < StateVector.Length; i++)
            {
                StateVector[i] /= norm;
            }
        }
    }

    /// <summary>
    /// Create a copy of the state.
    /// </summary>
    /// <returns>A new PhotonicState with copied state vector</returns>
    public PhotonicState Copy()
    {
        return new PhotonicState(StateVector, normalize: false);
    }

    /// <summary>
    /// Create a |0...0⟩ state with n qubits.
    /// </summary>
    /// <param name="numQubits">Number of qubits</param>
    /// <returns>PhotonicState in |0...0⟩ state</returns>
    public static PhotonicState ZeroState(int numQubits)
    {
        int size = 1 << numQubits;
        var stateVector = new Complex[size];
        stateVector[0] = Complex.One;
        return new PhotonicState(stateVector, normalize: false);
    }

    /// <summary>
    /// Create a |1...1⟩ state with n qubits.
    /// </summary>
    /// <param name="numQubits">Number of qubits</param>
    /// <returns>PhotonicState in |1...1⟩ state</returns>
    public static PhotonicState OneState(int numQubits)
    {
        int size = 1 << numQubits;
        var stateVector = new Complex[size];
        stateVector[size - 1] = Complex.One;
        return new PhotonicState(stateVector, normalize: false);
    }

    /// <summary>
    /// Create a computational basis state |i⟩.
    /// </summary>
    /// <param name="numQubits">Number of qubits</param>
    /// <param name="basisIndex">Index of the basis state (0 to 2^n - 1)</param>
    /// <returns>PhotonicState in basis state |i⟩</returns>
    public static PhotonicState BasisState(int numQubits, int basisIndex)
    {
        int size = 1 << numQubits;
        if (basisIndex < 0 || basisIndex >= size)
        {
            throw new ArgumentException($"Basis index must be between 0 and {size - 1}");
        }

        var stateVector = new Complex[size];
        stateVector[basisIndex] = Complex.One;
        return new PhotonicState(stateVector, normalize: false);
    }

    /// <summary>
    /// Create an equal superposition state (|0⟩ + |1⟩)/√2 ⊗ ... ⊗ (|0⟩ + |1⟩)/√2.
    /// </summary>
    /// <param name="numQubits">Number of qubits</param>
    /// <returns>PhotonicState in equal superposition</returns>
    public static PhotonicState Superposition(int numQubits)
    {
        int size = 1 << numQubits;
        double amplitude = 1.0 / Math.Sqrt(size);
        var stateVector = new Complex[size];
        for (int i = 0; i < size; i++)
        {
            stateVector[i] = new Complex(amplitude, 0);
        }
        return new PhotonicState(stateVector, normalize: false);
    }

    /// <summary>
    /// Compute the density matrix ρ = |ψ⟩⟨ψ| for this pure state.
    /// </summary>
    /// <returns>Density matrix as a 2D complex array</returns>
    public Complex[,] DensityMatrix()
    {
        int size = StateVector.Length;
        var rho = new Complex[size, size];
        for (int i = 0; i < size; i++)
        {
            for (int j = 0; j < size; j++)
            {
                rho[i, j] = StateVector[i] * Complex.Conjugate(StateVector[j]);
            }
        }
        return rho;
    }

    /// <summary>
    /// Check if the state is normalized.
    /// </summary>
    /// <param name="tolerance">Numerical tolerance for the norm check</param>
    /// <returns>True if ||ψ|| ≈ 1</returns>
    public bool IsNormalized(double tolerance = 1e-10)
    {
        double norm = Math.Sqrt(StateVector.Sum(c => c.Real * c.Real + c.Imaginary * c.Imaginary));
        return Math.Abs(norm - 1.0) < tolerance;
    }

    /// <summary>
    /// Compute the probability of measuring a specific basis state.
    /// </summary>
    /// <param name="basisIndex">Index of the basis state</param>
    /// <returns>Probability P(|i⟩) = |⟨i|ψ⟩|²</returns>
    public double Probability(int basisIndex)
    {
        if (basisIndex < 0 || basisIndex >= StateVector.Length)
        {
            throw new ArgumentException($"Basis index must be between 0 and {StateVector.Length - 1}");
        }

        var c = StateVector[basisIndex];
        return c.Real * c.Real + c.Imaginary * c.Imaginary;
    }

    /// <summary>
    /// Compute probabilities for all basis states.
    /// </summary>
    /// <returns>Array of probabilities for each basis state</returns>
    public double[] Probabilities()
    {
        return StateVector.Select(c => c.Real * c.Real + c.Imaginary * c.Imaginary).ToArray();
    }

    /// <summary>
    /// Compute inner product ⟨φ|ψ⟩ with another state.
    /// </summary>
    /// <param name="other">Another PhotonicState</param>
    /// <returns>Complex inner product</returns>
    public Complex InnerProduct(PhotonicState other)
    {
        if (NumQubits != other.NumQubits)
        {
            throw new ArgumentException("States must have the same number of qubits");
        }

        Complex sum = Complex.Zero;
        for (int i = 0; i < StateVector.Length; i++)
        {
            sum += Complex.Conjugate(other.StateVector[i]) * StateVector[i];
        }
        return sum;
    }

    /// <summary>
    /// Compute fidelity |⟨φ|ψ⟩|² with another state.
    /// </summary>
    /// <param name="other">Another PhotonicState</param>
    /// <returns>Fidelity value between 0 and 1</returns>
    public double Fidelity(PhotonicState other)
    {
        var ip = InnerProduct(other);
        return ip.Real * ip.Real + ip.Imaginary * ip.Imaginary;
    }

    /// <summary>
    /// Compute expectation value ⟨ψ|O|ψ⟩ of an operator.
    /// </summary>
    /// <param name="operatorMatrix">Operator matrix</param>
    /// <returns>Expectation value</returns>
    public Complex ExpectationValue(Complex[,] operatorMatrix)
    {
        int size = StateVector.Length;
        if (operatorMatrix.GetLength(0) != size || operatorMatrix.GetLength(1) != size)
        {
            throw new ArgumentException("Operator matrix dimensions must match state vector size");
        }

        // Compute O|ψ⟩
        var result = new Complex[size];
        for (int i = 0; i < size; i++)
        {
            result[i] = Complex.Zero;
            for (int j = 0; j < size; j++)
            {
                result[i] += operatorMatrix[i, j] * StateVector[j];
            }
        }

        // Compute ⟨ψ|O|ψ⟩
        Complex expectation = Complex.Zero;
        for (int i = 0; i < size; i++)
        {
            expectation += Complex.Conjugate(StateVector[i]) * result[i];
        }
        return expectation;
    }

    /// <summary>
    /// Compute partial trace over specified qubits.
    /// </summary>
    /// <param name="tracedQubits">List of qubit indices to trace out</param>
    /// <returns>Reduced density matrix</returns>
    public Complex[,] PartialTrace(List<int> tracedQubits)
    {
        // Get full density matrix
        var rho = DensityMatrix();

        // Determine kept qubits
        var keptQubits = Enumerable.Range(0, NumQubits)
            .Where(i => !tracedQubits.Contains(i))
            .ToList();

        if (keptQubits.Count == 0)
        {
            // All qubits traced out, return scalar
            Complex trace = Complex.Zero;
            for (int i = 0; i < rho.GetLength(0); i++)
            {
                trace += rho[i, i];
            }
            return new Complex[,] { { trace } };
        }

        // Dimensions
        int dimKept = 1 << keptQubits.Count;
        int dimTraced = 1 << tracedQubits.Count;

        // Initialize reduced density matrix
        var rhoReduced = new Complex[dimKept, dimKept];

        // Sum over traced qubits
        for (int i = 0; i < dimKept; i++)
        {
            for (int j = 0; j < dimKept; j++)
            {
                // Map reduced indices to full indices
                for (int k = 0; k < dimTraced; k++)
                {
                    // Build full indices
                    int idxI = 0;
                    int idxJ = 0;

                    // Insert kept qubit values
                    for (int pos = 0; pos < keptQubits.Count; pos++)
                    {
                        int qubit = keptQubits[pos];
                        int bitI = (i >> pos) & 1;
                        int bitJ = (j >> pos) & 1;
                        idxI |= (bitI << qubit);
                        idxJ |= (bitJ << qubit);
                    }

                    // Insert traced qubit values (same for both)
                    for (int pos = 0; pos < tracedQubits.Count; pos++)
                    {
                        int qubit = tracedQubits[pos];
                        int bit = (k >> pos) & 1;
                        idxI |= (bit << qubit);
                        idxJ |= (bit << qubit);
                    }

                    rhoReduced[i, j] += rho[idxI, idxJ];
                }
            }
        }

        return rhoReduced;
    }

    /// <summary>
    /// String representation of the state in Dirac notation.
    /// </summary>
    /// <returns>String representation</returns>
    public override string ToString()
    {
        var result = new List<string>();
        for (int i = 0; i < StateVector.Length; i++)
        {
            if (StateVector[i].Magnitude > 1e-10)
            {
                string basis = Convert.ToString(i, 2).PadLeft(NumQubits, '0');
                if (Math.Abs(StateVector[i].Imaginary) < 1e-10)
                {
                    result.Add($"{StateVector[i].Real:+0.####}|{basis}⟩");
                }
                else
                {
                    result.Add($"({StateVector[i].Real:+0.####}{StateVector[i].Imaginary:+0.####}i)|{basis}⟩");
                }
            }
        }
        return result.Count > 0 ? string.Join(" + ", result).Replace("+-", "-") : "0";
    }
}
